diff --git a/radio/file_contexts b/radio/file_contexts
index 6146f4d..b0c4a3a 100644
--- a/radio/file_contexts
+++ b/radio/file_contexts
@@ -41,3 +41,5 @@
 /dev/umts_rfs0                                                              u:object_r:radio_device:s0
 /dev/umts_dm0                                                               u:object_r:radio_device:s0
 /dev/umts_router                                                            u:object_r:radio_device:s0
+
+/vendor/bin/write_at_commands\.sh   u:object_r:write_at_cmd_exec:s0
diff --git a/radio/write_at_cmd.te b/radio/write_at_cmd.te
new file mode 100644
index 0000000..fe821b8
--- /dev/null
+++ b/radio/write_at_cmd.te
@@ -0,0 +1,10 @@
+type write_at_cmd, domain;
+type write_at_cmd_exec, exec_type, vendor_file_type, file_type;
+init_daemon_domain(write_at_cmd)
+allow write_at_cmd radio_device:chr_file rw_file_perms;
+allow write_at_cmd logd:fd use;
+allow write_at_cmd logd_socket:sock_file write;
+allow shell write_at_cmd_exec:file getattr;
+allow write_at_cmd shell_exec:file execute_no_trans;
+allow write_at_cmd toolbox_exec:file { read getattr execute open };
+allow write_at_cmd radio_device:chr_file ioctl;
